### PR TITLE
Refactor Notice component to functional Typescript

### PR DESCRIPTION
**Pull request by cpapazoglou**
#### Changes proposed in this Pull Request

In an effort to decouple the `Notices` component from calypso, I am also rewriting the needed components to TS functional components hoping that this will reduce some regressions.

**Todo**

- [ ] Fix the tests https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/components/notice/test/index.js#L14

#### Testing instructions

* Inspect /devdocs/design/notice in this live branch and in https://wpcalypso.wordpress.com/devdocs/design/
* Go to posts screen, delete one post and then undo.
* Check for regressions

Related to #51276

### CODE REVIEW

File: client/components/notice/index.tsx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+		}
+
+		return () => {
+			if ( dismissTimeout ) {
+				clearTimeout( dismissTimeout );
+			}
+		};
+	}, [ duration, onDismissClick ] );
+
+	// showDismiss = ! isCompact, // by default, show on normal notices, don't show on compact ones
```
Comment (sarayourfriend) [inline client/components/notice/index.tsx]: Did you mean to leave this?

File: client/components/notice/index.tsx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	children,
+	status = null,
+	icon = null,
+	isCompact = false,
+	isLoading = false,
+	showDismiss = ! isCompact,
+	text = null,
+} ) => {
+	useEffect( () => {
+		let dismissTimeout: TimerHandle | undefined = undefined;
```
Comment (sarayourfriend) [inline client/components/notice/index.tsx]: `dismissTimeout` won't be persisted between effect calls. Did you mean to create a ref outside the effect to store this in and change in the effect?

File: client/components/notice/index.tsx (hunk 1 above)
Comment (cpapazoglou) [reply client/components/notice/index.tsx]: Removed, thanks!

File: client/components/notice/index.tsx (hunk 2 above)
Comment (cpapazoglou) [reply client/components/notice/index.tsx]: Thanks @sarayourfriend for chiming in!

That's a good question. Initially I thought that not persisting it would be enough, I changed my mind though f3518bc

File: client/components/notice/index.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
+/**
+ * External dependencies
+ */
+
```
Comment (tyxla) [inline client/components/notice/index.tsx]: Nit: we can remove this empty line:
```suggestion
```

File: client/components/notice/index.tsx [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+import React, { useEffect, FunctionComponent, useRef } from 'react';
+import classnames from 'classnames';
+import { localize } from 'i18n-calypso';
+import Gridicon from 'calypso/components/gridicon';
```
Comment (tyxla) [inline client/components/notice/index.tsx]: Nit: Let's move the `Gridicon` import to the `Internal dependencies` block below.

File: client/components/notice/index.tsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+/**
+ * Internal dependencies
+ */
+import { TimerHandle } from 'calypso/types';
+/**
```
Comment (tyxla) [inline client/components/notice/index.tsx:15]: Nit: We could leave an empty line here to separate those blocks properly:

```suggestion

/**
```

File: client/components/notice/index.tsx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+ * Style dependencies
+ */
+import './style.scss';
+
+/**
+ * Module constants
+ */
+const GRIDICONS_WITH_DROP = [
+	'add',
```
Comment (tyxla) [inline client/components/notice/index.tsx:24]: Nit: Have you considered sorting these alphabetically? I've noticed that this usually makes unordered lists more readable.

File: client/components/notice/index.tsx [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+	'pause',
+	'play',
+	'spam',
+];
+
+/**
+ * Constants
+ */
+enum Status {
+	Error = 'is-error',
```
Comment (tyxla) [inline client/components/notice/index.tsx]: Just like above, should we sort these alphabetically?

File: client/components/notice/index.tsx [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	Info = 'is-info',
+	Success = 'is-success',
+	Warning = 'is-warning',
+	Plain = 'is-plain',
+}
+
+interface Props {
+	className?: string;
+	duration?: number;
+	onDismissClick: () => void;
```
Comment (tyxla) [inline client/components/notice/index.tsx:38]: What I've observed during the refactoring away from Lodash's `noop` is that it tends to be cleaner if we don't provide a default prop to the function props, but then check if it's provided before using it. I think here it would be a good opportunity for refactoring, as if `onDismissClick` isn't set, we won't call a noop function unnecessarily, and all the related `setTimeout` / `clearTimeout` that are related to that.

File: client/components/notice/index.tsx [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	translate: ( x: string ) => string;
```
Comment (tyxla) [inline client/components/notice/index.tsx]: Should this use `LocalizeProps[ 'translate' ]` from `'i18n-calypso'` instead?

File: client/components/notice/index.tsx [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+	'notice',
+	'pause',
+	'play',
+	'spam',
+];
+
+/**
+ * Constants
+ */
+enum Status {
```
Comment (tyxla) [inline client/components/notice/index.tsx]: We already have existing types for the notice statuses - [see here](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/notices/types.ts#L16). 

Maybe they would be worth moving to a more generic place now that they'll be used for more than just state?

Also, a side note here is that I'd prefer using [union types instead of enums](https://blog.bam.tech/developer-news/should-you-use-enums-or-union-types-in-typescript) for those, but this is already done for `NoticeStatus` as you can see.

File: client/components/notice/index.tsx [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+	Error = 'is-error',
+	Info = 'is-info',
+	Success = 'is-success',
+	Warning = 'is-warning',
+	Plain = 'is-plain',
+}
+
+interface Props {
+	className?: string;
+	duration?: number;
```
Comment (tyxla) [inline client/components/notice/index.tsx:37]: I think `duration` can also be `null`. This is how we declared it in https://github.com/Automattic/wp-calypso/blob/trunk/client/state/notices/types.ts#L29 as well. Is it possible we're not handling it properly here or there?

File: client/components/notice/index.tsx [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+	onDismissClick: () => void;
+	translate: ( x: string ) => string;
+	children: React.ReactNode;
+	status?: Status;
+	icon?: string | null;
+	isCompact?: boolean;
+	isLoading?: boolean;
+	showDismiss?: boolean;
+	text?: string | React.ReactNode;
```
Comment (tyxla) [inline client/components/notice/index.tsx]: This one looks good, but we already have a `NoticeText` [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/notices/types.ts#L23).

File: client/components/notice/index.tsx [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+const getIcon = ( status: Status ): string => {
+	let icon: string;
```
Comment (tyxla) [inline client/components/notice/index.tsx]: Do we really need the `icon` variable? Can we just `return 'info'` and so on in each `case` statement? I know this was like this before, but since we're touching it, it's a good opportunity to improve it.

File: client/components/notice/index.tsx [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+		default:
+			icon = 'info';
+			break;
+	}
+
+	return icon;
+};
+
+const Notice: FunctionComponent< Props > = ( {
+	className = '',
```
Comment (tyxla) [inline client/components/notice/index.tsx]: It's so weird to pass an empty string as default here. Won't an `undefined` work the same way when evaluated by `classNames` below?

File: client/components/notice/index.tsx [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+} ) => {
+	const dismissTimeout = useRef< TimerHandle | null >( null );
+
+	useEffect( () => {
+		if ( dismissTimeout.current ) {
+			clearTimeout( dismissTimeout.current );
+		}
+
+		if ( duration > 0 ) {
+			dismissTimeout.current = setTimeout( onDismissClick, duration );
```
Comment (tyxla) [inline client/components/notice/index.tsx:82]: Should we be doing these only if `onDismissClick` is defined and not a noop function?

File: client/components/notice/notice-action.tsx [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * Style dependencies
+ */
+import './style.scss';
+
+interface Props {
+	'aria-label': string;
+	href: string;
+	onClick: () => void;
```
Comment (tyxla) [inline client/components/notice/notice-action.tsx:15]: Ideally, we shouldn't default to noop functions if it's not necessary. It seems like defaulting to nothing will do the same job, but will spare us from declaring an extra noop function and executing it needlessly. That being said, should `onClick` be typed as an optional prop(with `?`)?

File: client/components/notice/notice-action.tsx [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+	href,
+	onClick,
+	children,
+	external = false,
+	icon,
+} ) => {
+	const attributes = {
+		'aria-label': ariaLabel,
+		className: 'notice__action',
+		href: href,
```
Comment (tyxla) [inline client/components/notice/notice-action.tsx]: We could shorten this to:

```suggestion
		href,
```

File: client/components/notice/notice-action.tsx [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+ */
+import './style.scss';
+
+interface Props {
+	'aria-label': string;
+	href: string;
+	onClick: () => void;
+	children: ReactNode;
+	external?: boolean;
+	icon?: string;
```
Comment (tyxla) [inline client/components/notice/notice-action.tsx:18]: Can `icon` be `null` too?

File: client/components/notice/notice-action.tsx [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+} ) => {
+	const attributes = {
+		'aria-label': ariaLabel,
+		className: 'notice__action',
+		href: href,
+		onClick: onClick,
+	};
+
+	if ( external ) {
+		attributes.target = '_blank';
```
Comment (tyxla) [inline client/components/notice/notice-action.tsx:37]: Now that we're touching this one, it would be ideal if we avoid mutations, even if they seem pretty safe here.

File: client/my-sites/domains/components/mapping-instructions/index.jsx [hunk 20]
```diff
@@ -16,7 +16,7 @@ import {
 } from 'calypso/lib/url/support';
 import { WPCOM_DEFAULT_NAMESERVERS } from 'calypso/state/domains/nameservers/constants';
 import FoldableFAQ from 'calypso/components/foldable-faq';
-import { Notice } from 'calypso/components/notice';
+import Notice from 'calypso/components/notice';
```
Comment (tyxla) [inline client/my-sites/domains/components/mapping-instructions/index.jsx:19]: 😅  Nice fix!

**Review by tyxla — COMMENTED** [review]
Nice work, thanks for working on this 👏 

I've left some feedback, the highlight is that we already have some overlapping types created for the notice actions, and it would make sense to share them between the state and the component.

File: client/components/notice/index.tsx [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+	isLoading = false,
+	showDismiss = ! isCompact, // by default, show on normal notices, don't show on compact ones
+	text,
+} ) => {
+	const dismissTimeout = useRef< TimerHandle | null >( null );
+
+	useEffect( () => {
+		if ( dismissTimeout.current ) {
+			clearTimeout( dismissTimeout.current );
+		}
```
Comment (jsnajdr) [inline client/components/notice/index.tsx]: This `clearTimeout` call should not be necessary. The one in the effect cleanup callback should be sufficient to always clear the timeout when needed.

File: client/components/notice/index.tsx [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+					aria-label={ translate( 'Dismiss' ) }
+				>
+					<Gridicon icon="cross" size={ 24 } />
+				</button>
+			) }
+		</div>
+	);
+};
+
+export default localize( Notice );
```
Comment (jsnajdr) [inline client/components/notice/index.tsx]: If the `Notice` component is functional, let's use the `useTranslate` hook instead of `localize`. It will also save us some work with declaring the `translate` prop.

File: client/components/notice/index.tsx (hunk 6 above)
Comment (cpapazoglou) [reply client/components/notice/index.tsx:24]: Haven't till now! Very good idea, aren't these already sorted though?

File: client/components/notice/index.tsx (hunk 7 above)
Comment (cpapazoglou) [reply client/components/notice/index.tsx]: Good idea, done!

File: client/components/notice/index.tsx (hunk 10 above)
Comment (cpapazoglou) [reply client/components/notice/index.tsx]: Thanks for the heads up. This is an interesting approach to union types vs enums! 

I have refactored it: 8bfc341, c11de0e

File: client/components/notice/notice-action.tsx (hunk 17 above)
Comment (cpapazoglou) [reply client/components/notice/notice-action.tsx]: Good idea!

Comment (cpapazoglou) [reply client/components/notice/notice-action.tsx]: Good idea!

File: client/components/notice/index.tsx (hunk 12 above)
Comment (cpapazoglou) [reply client/components/notice/index.tsx]: Good point! 48097a9

File: client/components/notice/index.tsx (hunk 14 above)
Comment (cpapazoglou) [reply client/components/notice/index.tsx]: Good point, https://github.com/Automattic/wp-calypso/pull/51686/commits/48097a9e14d33a5dcba11bc748501f651fc8b843

File: client/components/notice/index.tsx (hunk 13 above)
Comment (cpapazoglou) [reply client/components/notice/index.tsx]: Sure d18ce42!

File: client/components/notice/index.tsx (hunk 6 above)
Comment (tyxla) [reply client/components/notice/index.tsx:24]: Indeed, they totally are 👍

File: client/components/notice/index.tsx (hunk 11 above)
Comment (cpapazoglou) [reply client/components/notice/index.tsx:37]: I have searched all over Calypso and couldn't find a reference with `null` duration b05d68e

File: client/components/notice/index.tsx (hunk 21 above)
Comment (cpapazoglou) [reply client/components/notice/index.tsx]: Good point cf9d61e!

File: client/components/notice/index.tsx (hunk 11 above)
Comment (tyxla) [reply client/components/notice/index.tsx:37]: Might be a remnant of the pre-Redux era. If we never provide a `null` for duration it would be nice to update the [notices  readme](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/notices/README.md#L43) and their [types](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/notices/types.ts#L29). Ideally, we should reuse as much as possible from there, too.

File: client/components/notice/index.tsx (hunk 15 above)
Comment (jsnajdr) [reply client/components/notice/index.tsx:82]: The `dismissTimeout` doesn't need to be stored in a ref. It can be a local variable in the `useEffect` callback scope.

File: client/components/notice/index.tsx [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+
+	useEffect( () => {
+		if ( duration > 0 ) {
+			dismissTimeout.current = setTimeout( onDismissClick, duration );
+		}
+
+		return () => {
+			if ( dismissTimeout.current ) {
+				clearTimeout( dismissTimeout.current );
+			}
```
Comment (jsnajdr) [inline client/components/notice/index.tsx:88]: This code could call `clearTimeout` twice on the same timer if we did a prop transition from `duration={ 1000 }` to `duration={ 0 }` and then to `duration={ 1000 }`.

File: client/components/notice/index.tsx [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+		};
+	}, [ duration, onDismissClick ] );
```
Comment (jsnajdr) [inline client/components/notice/index.tsx:90]: You probably don't want to reset the timeout when `onDismissClick` callback changes. Better to store it in a ref:
```js
useEffect( () => {
  dismissCallback.current = onDismissClick;
}, [ onDismissClick ] );

useEffect( () => {
  let id = setTimeout( () => dismissCallback.current(), duration );
  return () => clearTimeout( id );
}, [ duration ] );
```
Dan Abramov has a nice article about this where he explains everything:
https://overreacted.io/making-setinterval-declarative-with-react-hooks/

File: client/components/notice/index.tsx [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+	className?: string;
+	duration?: number;
+	onDismissClick: () => void;
+	children: React.ReactNode;
+	status?: NoticeStatus;
+	icon?: string | null;
+	isCompact?: boolean;
+	isLoading?: boolean;
+	showDismiss?: boolean;
+	text?: NoticeText;
```
Comment (jsnajdr) [inline client/components/notice/index.tsx:45]: This type can be simply a `React.ReactNode`. It already includes `string` as one of possible values.

File: client/components/notice/index.tsx [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+const getIcon = ( status: NoticeStatus ): string => {
```
Comment (jsnajdr) [inline client/components/notice/index.tsx:48]: `status` can be also `undefined`. Or you can default the `status` prop to `is-plain` in the component body.
